libs/editor/egui_editor/src/android/java_ffi.rs

```diff
@@ -0,0 +1,344 @@
+//         obj.editor.buffer.current.cursor.selection.end() + (n as usize)
+//     )
+//         .into();
+//
+//     let buffer = &obj.editor.buffer.current;
+//     let text = cursor.selection_text(buffer);
+//
+//     env
+//         .new_string(text)
+//         .expect("Couldn't create JString from rust string!")
+//         .into_raw()
+// }
+
+#[no_mangle]
+pub extern "system" fn Java_app_lockbook_egui_1editor_EGUIEditor_deleteSurroundingText(
+    _env: JNIEnv, _: JClass, obj: jlong, before_length: jint, after_length: jint
+) {
+    let obj = unsafe { &mut *(obj as *mut WgpuEditor) };
+
+    let current_cursor = &obj.editor.buffer.current.cursor;
+
+    obj.editor.custom_events.push(Modification::Replace {
+        region: Region::BetweenLocations {
+            start: Location::DocCharOffset(current_cursor.selection.start() + (before_length as usize)),
+            end: Location::DocCharOffset(current_cursor.selection.start())
+        },
+        text: "".to_string()
+    });
+
+    obj.editor.custom_events.push(Modification::Replace {
+        region: Region::BetweenLocations {
+            start: Location::DocCharOffset(current_cursor.selection.end() + (after_length as usize)),
+            end: Location::DocCharOffset(current_cursor.selection.end())
+        },
+        text: "".to_string()
+    });
+}
+
+#[no_mangle]
+pub extern "system" fn Java_app_lockbook_egui_1editor_EGUIEditor_getAllText(
+    env: JNIEnv, _: JClass, obj: jlong
+) -> jstring {
+    let obj = unsafe { &mut *(obj as *mut WgpuEditor) };
+
+    env
+        .new_string(&obj.editor.buffer.current.text)
+        .expect("Couldn't create JString from rust string!")
+        .into_raw()
+}
+
+#[no_mangle]
+pub extern "system" fn Java_app_lockbook_egui_1editor_EGUIEditor_setSelection(
+    _env: JNIEnv, _: JClass, obj: jlong, start: jint, end: jint
+) {
+    let obj = unsafe { &mut *(obj as *mut WgpuEditor) };
+
+    obj.editor.custom_events.push(Modification::Select {
+        region: Region::BetweenLocations {
+            start: Location::DocCharOffset(DocCharOffset(start as usize)),
+            end: Location::DocCharOffset(DocCharOffset(end as usize))
+        }
+    });
+}
+
+#[no_mangle]
+pub extern "system" fn Java_app_lockbook_egui_1editor_EGUIEditor_sendKeyEvent(
+    _env: JNIEnv, _: JClass, obj: jlong, key_code: jint, pressed: jboolean, alt: jboolean, ctrl: jboolean, shift: jboolean
+) {
+    let obj = unsafe { &mut *(obj as *mut WgpuEditor) };
+
+    let modifiers = egui::Modifiers {
+        alt: alt == 1,
+        ctrl: ctrl == 1,
+        shift: shift == 1,
+        mac_cmd: false,
+        command: false
+    };
+
+    obj.raw_input.modifiers = modifiers;
+
+    let Some(key) = AndroidKeys::from(key_code) else {
+        return
+    };
+
+    if let Some(key) = key.egui_key() {
+        obj.raw_input
+            .events
+            .push(Event::Key { key, pressed: pressed == 1, repeat: false, modifiers });
+    }
```

Comment:
may have to detect more situations here https://github.com/lockbook/lockbook/blob/65dea5db545af7b583e088c7e79e700e9feea6dc/libs/editor/egui_editor/src/apple/macos_ffi.rs#L22-L41

Reply:
Added, thanks.